### ISSUE TITLE
TypeScript ignoring checks for deeply nested computed object types

### ISSUE THREAD

**Issue by sinclairzx81** [microsoft/TypeScript#56138]
### 🔎 Search Terms

inference, deeply nested, no error, type evaluation, typebox, hkt

### 🕗 Version & Regression Information

Hi, 

I've come across some unusual behavior where TypeScript seems to be skipping checks for certain kinds of computed types. This seems to be specific to deeply nested object types that are computed via intersection.  I note that the language service is reporting the correct type information in editor, however the compiler seems to be skipping checks on the reported type. This behavior can be observed using the TypeBox library, specifically for object types at least 2 levels deep.

- This is the behavior in every version I tried (5.1.6 onwards)

### ⏯ Playground Link

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/PTAEFpK6dv4Yp8BQJQHkBOBLA5jgHYCGANqAMqE4AONApgC5pjJvsefQr0AeNAeyyNQjAJ4NQASUI0AriIC8lRsUY4AxgB5xDAQDNpshQD4e-ISI0DCAZxEz5S0ABUJ9AHQYARgCt6GowAFADeKKCgpPQAbvSkAIwAXK7uXn4BwWEREVGxpABMyW4Maf6BoeHZEfoCAkWpFIy4hHhBAJQANJURAL5tlX0og+aCwqLumApOoMqNapo67gaTjE5mfKNWNvYr08rFnj5lmZW5cUkpJUcZFVVnBfVX6eVZVaA1dZeejc2tnd3ZbzELCPb5NIh-LpVQa9frDFiuAAW9HeAlIpAEAHcIe85IRAjhtqBbIiBHJSAATUBkchYeibUD0LBYIS2VFYUTI0B0xhyLCEUDRMhyegdUCkzExJmclH6HBYHYUgT0NmEASMDwoFD6PEEmygGgs7xRAC28w0ADFdeobPEgqzko4FABtAC6bWSGCmLtdoFePL5AtZAG5QOgtJBOTg2UqVaA1SImSyOUMtTr8TaBYaBMb6Gb1JbrYTCPktC5GbxGPRCBS2V7Vj6TPaFY7jIw3R7XH7KgH+aAQ2GwBHwKAPGPQN4FFGY8rbKntUX9dnc-nNFaM8WAMzN2zJIK6ejLJ0a+zmtodz3e9u+-1MQP9hWh8ORsceak14kCUCxtmMRHR+cES4YCQNAhEDgAIQEXgjFIIh6CpCC4ixIDQLQ9DEC1dAMJw3DIARABVdQ4PEVC8PIkCRkscZJAAUSFUg5DUegyxMGYuz4KsazZIh9GlDBQAAfj9UBnQAaVAIhQAAa3oMRlgwV1PXE30elAZJCClLAyIo3T2ARCgxBNbw0TnbC9Is5AqLGCkAlIYEUWsOwRAAJXoYglUIUgxGSPEcAARxFYkjJM0hg2skRbI0ey6VAJydgwGhMzIXzqEClFbBCtFwo2aiopixzthEAAJIhGFSgKgsy4zsoi787IcuKitAMSiApCr0uCmqwp0yy+pgcCKA0ZEzV6-rxrqsqmX0YgNBRFxWo-V5xLapTiXBFpU1ysYpqwGa5tcIaRuICsuNrVxFqpZa3I8mxvNdQTknsX5KmdRLktIB6no2vBXtKwhGC+9aXoiGhgWIE1d1APFpLVTFCDdSpTwLVLYaxQhU3M8bsfABEAFkBApHA5SZMzWBx7G6oPVwbs87z3uLMgy1O6tzpcI682INj9gZmwmZcNiADIafcumxFYqmJhcWm7vF8tONZtl2eGznua7YWQlEmWvLENaAHJte8vXQB6SXJBcXmSFIZmFe4w6VbNNXyw10TLbIfW3dIY3Taxim+vA65AjGv3dMmgHptm+afhxW22Y5s1uwiFaa316OWj18KImRzRvt+TOaPoZI9eeiEM8x8mQ8sgPnmYX3K9D7aRGpw2xE9gAFFkGGEMQxLk2wbcrRXXA7gQu-UFU1c18TJIFWT5MMFw1pcFSWbt6XRdlst465oTQCCZexN9WOlc9ssVLY4SJI0rS2nU+MtJN505+WRezZRFuR7Hnu+4Hs6lc-pk49bCT1EhJKSz8F5LxXsfEWt0dZbwdjvYS+9oGDzXqfA+roL731iByZIYlb7X1wY-CBrhXRv0wElRmpAAHd17mIfu8s0Fs1oUAkB09wFyRflAw+q82YYO3tglBvCYHrzgd5M+h9sGaWIfgwhODpQ9CflwyBFC3KBXlAhVh38GG-yHi4VhOAJ7sVIRgE0OBGBljFC3dundAE6MYWxAAPqAD+di6E-wFqAFxtjR72PoY49YFgxjU0MSqNyFI5BzSwHoteYTbDWL4WyNy1gsAUi0KQ4ghAxBihhnDQgJg1b0WFMxLQQRKgty0G3YE6gmZt00NJLQLlrEbx1r4r+ATWKFKFhU1pEj6kaEac01xfSxDaM6QLQpoBBaVGqcIHAdSGlNLFO0-xnjunTN6Rouk6SBlDJadsrR7jxATO6SgNoQSGShOOUY2wESoksSYX-YeNyVRijbkk6GhA0bwzdGreJ9zolWMTqAmeMkVFkOSHMAskjXTvLMD0S51FqYGOOTo9iJcWjePjHIYyTIKGor8fMuMygUlCHSYSjpckxTK2Okina4c9qR1cIHSxTz9HxPYpSwBty2KiO3iC5OFIPY1z1kjVQKNQAAoQg84Ff5ozOj1mDLAENbB6ywZUA8RcczHDFaDV5UMXDlwgPXCyBkJULAFsHU1OEKHQstZ82lnN3mfLyejN07E-nsX3tMkSyrVXJDbqGPoirs4aHVdam16FwITAAHIQxVGDOakao1gXyo1EgJpE3MoOCC9MeoBSstiSwg1TYdUZCKJ2C2NcJbVEXAKNOfwiiNtTEAA)

### 💻 Code

```ts
import { Static, Type } from "@sinclair/typebox"

export type Input = Static<typeof Input>
export const Input = Type.Object({
  level1: Type.Object({
    level2: Type.Object({
      foo: Type.String(),
    })
  })
})

export type Output = Static<typeof Output>
export const Output = Type.Object({
  level1: Type.Object({
    level2: Type.Object({
      foo: Type.String(),
      bar: Type.String(),
    })
  })
})

// The following functions should all report errors for the return value, however the first does not.

function problematicFunction1(ors: Input[]): Output[] {
  return ors; // <-- this does not error 
}

function problematicFunction2<T extends Output[]>(ors: Input[]): T {
  return ors; // <-- ... but this does
}

function problematicFunction3(ors: (typeof Input.static)[]): Output[] {
  return ors; // <-- ... and so does this
}
```

### 🙁 Actual behavior

First function does not report an error

### 🙂 Expected behavior

All functions should report errors

### Additional information about the issue

I get the feeling this may have something to do with TypeBox's implementation of `Static<T>` and where TypeScript may be implementing some optimizations around certain forms of computed / evaluated types (but not sure). TypeBox currently implements `Static<T>` as follows.

```ts
type Static<T extends TSchema, P extends unknown[] = []> = (T & { params: P; })["static"]
```
By removing the intersection of P, TypeScript correctly asserts the `problematicFunction1` return type.
```ts
type Static<T extends TSchema, P extends unknown[] = []> = T["static"] // this is fine
```

**Comment by RyanCavanaugh** [comment]
Please inline the relevant parts of `typebox`

**Comment by sinclairzx81** [comment]
@RyanCavanaugh See below for the relevant parts of TypeBox inlined (Issue TSP link also updated)

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/PTAEFpK6dv4Yp8BQJQHkBOBLA5gHYCGANqAMoE4AO1ApgC5pjKtvsfQp0Ae1A9lgagGAT3qgAkgWoBXYQF4KDIgxwBjADxj6-AGZSZ8gHzc+g4ev4EAzsOlzFoACri6AOgwAjAFZ11DAAUAN4ooKAkdABudCQAjABcLm6evv5BoeHhkTEkAExJrvSpfgEhYVnhevz8hSnkDLgEeIEAlAA0FeEAvq0VvSgDZgJCIm6Y8o6gSg2qGtpu+hMMjqa8I5bWdstTSkUe3qUZFTmxicnFh+nllaf5dZdpZZmVoNW1Fx4NTS0dXVleIhYB5fRo4ZptTqVAY9PpDdDOAAWdDe-BIJH4AHdwXg3rICAEcFtQDZEfxZCQACagUhkLB0DagOhYLCCGyorAiZGgekMWRYAigKKkWR0dqgMmY6LMrkovQ4LDbSn8Ojsgj8BjuFAoPT4wnWUDUVleSIAWzm6gAYnq1NY4oE2UkHPIANoAXVaSQwk1dbtAL15-MFbIA3KB0JpIFycOzlarQOrhMzWZzBtrdQTbYKjfwTXRzWorTaiQQ8ppnEyeAw6ARKezvStfcYHYqnUYGO7PS5-RVAwLQKHw2BI+BQO5x6AvPJo7GVTY0zriwac3mCxprZmSwBmFs2JKBHR0JbOzV2C2tTten3CAOMIMDxVhiNR8fuGm1kn8UBx9kMRExhdmAgTgQNA0CgP2AAhfgeEMEhwToalINiLEgLA9CMKQbV0Ew3C8MgICAFU1HgsQ0PwiiwOGCwxgkABRYUSFkVQ6HLYxpm7Xhq1rdlwT0GUMFAAB+f1QBdABpUBwVAABrOhRCWDA3S9CS-W6UAkgIaUsHIyi9PYIDyFEU0vDReccP0yzkGo0ZKX8EggRRKxbGEZwsCIWxqiwU0knxHAAEdRRJYzTJIEMbOEOz1Ac+lQGc7YACU6CIZUCBIURfKoQKURsEK0XC9YaKimKnK2YQMGoLNSEygKgtykz8oi797McuKytAAAJcEGBq7LgoasKmuK1r4uEcTwUpXq6rywaLKs+a4Ag8h1GRc1dIWjaCMK0ZuuZPQiHUFFltWogKqzdkXgAEhsFb8yIISkjsH4KmAAAqV6pGpLyZxJW7zVAV7gAqS6cEpB6STBZoXvelwcAYSIBwMP8Y1+k6AaB8I1HhuhwaenFoY+gARVV1FwSqS0Rn6brRwGKjsm6yazXHIbwAnQGJ-aKWERigu+5H2Wpu70bpuhOZIBhwY80Q2bongiFNagEZ5+MCxWnEqb+ohhfCXh5cV1VJYIaXwjej6zpLUh30TOYDW++kUowNLRG1nlkspR30vB0y0WSgg2fN6xLY862sw5UBMVwasPed2nwgjuG6Gjr2akiDyKhdHNqEelnlPfaXuia3asH2w6XHGj8Xgkibc7xqGC+24Qi5LlFnGOoWuJrOsKE1gPbHFZxy+pSukpS6x0rdZnnvCF1e9ICfs6nsSuoIBh54hxfqCBeW91AfEZPVTECHdCoz0LTL96xP2C7mzaNqAgBZfhKRweVmXMlhb82prDxcEfUvS2eJByyVm4l3Vumt2J7EAWxUAAAyX+bsx6iDYt-cYzg-5IOAR3HiLg27mkgd2eBwQxIYKdrnAA5KQ9K5DQD13MKMH+zhoEVmwWAvBRACEViIWJQBFDAE0Ovh-T+C0IJXACOtYRllUESG+DiS0ggCy9w4uQykLFwBqFNHQGhAAfUA5CNFaNALolRLEdF6LujgEgZjyGgwIOACxVijF6LJHYYgmjrG2PAC4hgbjDHGJoFEAALB46gUQABs1j+Q4EibgcA9J+L0gJH4vRCpolOPIbIWQoMPGxPicyGsh0Yk4HUfmRWpj0k+BsNYcAAgi7WPpA5NQMRwCVOqbUlezJ6l0DwLwMxIR1LwNrngVoBV6HCB-rI5oABhaw3EGB0QJE-HESilDkIAOxeDhtYgAHJshg1jNnECwKIaxgUNSIRqU0FQeYDlEBsHQMJwSnH9LgevHEIzC4dOLgdI6LNe7slYeydh-yewmxhkiFE5oeA4FNLIU0bzmgRBrHgP8LsoUABlkV-nBgQOFXhmThTBR9CFoBTTghhXChFuJIjNFRbHUl4JMW0sRDivFBK2YAEFXZ4ApECSsRpVQ2AppvBg1Z+z8ypSSMkFJqSq0RC7EVYqCCTxxIS8MMMuVeQLFTFmUryRUlJaoFaLstWqHBpMvA8jvKqF7mq02LhuTOTmUyRZz9EV83-ALXV9KnU1nma6nE5qWYzI6SvBZVg3V4FtWzElvqV6ksQjgLWP8PUoyGS7ONDB76JqIPsFVzRwp0MZE3H5uDdWAtwZrcUFqQWV0HhQi15C1Wnw0AvVVFRDxJHIUMptaYb6SP0qIp4TB+0DsotIlEVDRCAIAAqsnoEIUQ4l5I2CwVWTu7JnBzv4AutQqoCHEIklJQUckFIGGcLnZwqkQEboQaPJ25Z2HsREoEK94k-QVqYeTQOQC31umfaASSmltKtA0gmbStCXSnqWBeidoAp3bt3Uulda7QGbsQ8yPdNgD1iUktJaD57L3Xs-VOx9EDhKgFfcR9dOCv1VV-apADWkYiciSOJUDwGWOQYIy4N0cHZ3zsw8h0Qq6WE0bARhoQOB90cUPXhk98kYNEffTe2jzCn0UaoypkjiCH1-qYxBtjHHwNce6FBxThG4NJUCgqRCkmxDLpE6h29W7BNSZk0oHjGAyUMHLOKKdAmd1Ccc6J9iuiENuYcyh5wYXMDfuICQezwnQtrDGbRFE9npM2CSpSWQh0sDOdo5l1U-nVNdySlYLAlJNA8aluKPeB8CDGAIQxEULFNCBAqKRmdQI1CkE0DOjQMlNAJX87pgB8XSBJZC2xZr7FYFdfG8gwb6hhujfg0t6b0W5twIqD1qT-WVtrfFIFpDM2YvNd2+EazWT6TVaOyNsbNm7tbacxd0wrRUuMh-sV7LiE8usTE2hlwv3xQzrK+yBrl93QEN+zlgHBXnDijrce2SFneNJFmIWcsqkwemG6F9mijDXscXTbo3FJlmRwdc0F9z7IlAVcENVmnZ35L9yffx4dv2YF7Ahy4MRvm+IykIgBwiYHmNU4bulzAXPIvJZ55xcTm6BeaCF5yEXFGeNi845LtLjCOWUmfvR37HF8R2XlFpakujwFo10d7VOfspcltLgLkFFbgXxYuhUFKRuLaJci1l8GzgDe+5-b9ioZKCC-dZZTnS4QoXR80mynSRaaLO5biroHLmTd7F++xT97CTvDtrenetXaBfkJPioM+oA4f-cOn5mcLpyGb3cqaGw5D-3e8N3DP3ieXAh972HgPqoO1uC7bmI4lfwiZ0w1lwovY6DPcQvmvAx9BHATHYO9AWP5gxYkVv-CcHd9aCz7Rwvte+dQ8Pu6DiMOOKvtecQ1v28kgzrDL0ZvLb1Cd4P4f3CEE4wAAcvLKqJvIdH-v-hhCgMNLFL4jYOAS3OMC8BmPqIKJnnziznPvug6MOoUF2EwsOiglUEuIKBam0IUBammEAA)

**Comment by RyanCavanaugh** [comment]
WIP reduction
```ts
export type Input = Static<typeof Input>
export const Input = Type.Object({
  level1: Type.Object({
    level2: Type.Object({
      foo: Type.String(),
    })
  })
})

export type Output = Static<typeof Output>
export const Output = Type.Object({
  level1: Type.Object({
    level2: Type.Object({
      foo: Type.String(),
      bar: Type.String(),
    })
  })
})

function problematicFunction1(ors: Input): Output {
  return ors; // <-- this does not error
}
function f() {
  problematicFunction1(null as any);
}
f();
 
export type Evaluate<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;
export interface TSchema {
  params: unknown[];
  static: unknown;
}
interface HasStatic { static: unknown }
interface HasParams { params: unknown[] }
type RecordOfHasStatics = Record<string, HasStatic>;
export type PropertiesReduce<T extends RecordOfHasStatics, P extends unknown[]> = Evaluate<{ [K in keyof T]: Static<T[K], P> }>;
export interface TObject<T extends RecordOfHasStatics> extends HasParams {
  static: PropertiesReduce<T, this['params']>;
  properties: T;
}
export type Static<T extends HasStatic, P extends unknown[] = []> = (T & { params: P; })['static']

declare namespace Type {
  function Object<T extends RecordOfHasStatics>(object: T): TObject<T>;
  function String(): TSchema & { static: string };;
}
```

**Comment by RyanCavanaugh** [comment]
As expected, we're hitting a depth limit
```
{"pid":1,"tid":1,"ph":"I","cat":"checkTypes","ts":416867.20000207424,"name":"recursiveTypeRelatedTo_DepthLimit","s":"g","args":{"sourceId":236,"sourceIdStack":[165,222,236],"targetId":229,"targetIdStack":[198,215,229],"depth":3,"targetDepth":3}},
```
Source 236 is `Evaluate<T>`
165 is `Evaluate<T>`
222 is `Evaluate<T>`
236 is `Evaluate<T>`

Target 229 is `Evaluate<T>`
198 is `Evaluate<T>`
215 is `Evaluate<T>`
229 is ... `Evaluate<T>`

The source stack is
```ts
{ level1: { level2: { foo: string; }; }; } -> { level2: { foo: string; }; } -> { foo: string; }
```
and the target stack is
```ts
{ level1: { level2: { foo: string; bar: string; }; }; } -> { level2: { foo: string; bar: string; }; } -> { foo: string; bar: string; }
```
Since each of these types are instantiations of the same symbol `Evaluate<T>`, it appears that we might be stuck in an infinitely-expanding recursive type like this one, where each level of expansion creates new fresh types that seem like they might be worth looking into:
```ts
type RabbitHole<T, K> = {
  next: RabbitHole<T, { obj: K }>
}

declare let x: RabbitHole<string, number>;
let y = x.next.next.next.next.next.next.next;
```
So we need some way to tell apart OP's situation - where the types are actually converging, probably - from the infinite-descent case where they'll never converge.

**Comment by RyanCavanaugh** [comment]
The PR tanked xstate performance; we'll need something smarter.

Reduced and annotated example (apologies for renaming but these make more sense to me and avoid conflicting with certain terms we use internally):
```ts
type Input = GetType<typeof Input, []>
const Input = MakeObject({
    level1: MakeObject({
        level2: MakeObject({
            foo: MakeString(),
        })
    })
});

type Output = GetType<typeof Output, []>;
const Output = MakeObject({
    level1: MakeObject({
        level2: MakeObject({
            foo: MakeString(),
            bar: MakeString(),
        })
    })
});

function problematicFunction1(ors: Input): Output {
    // Should error
    return ors;
}

// Representational types have a 'myType' field that represents
// the TypeScript type of the object, e.g.: MakeString below
interface HasType { myType: unknown }
declare function MakeString(): { myType: string };

// Reads the 'myType' field out of an object. 'P' exists for
// reverse inference purposes when dealing with object types (see MakeObject below)
type GetType<T extends HasType, P> = (T & { objectProps: P; })['myType'];
// Mapped version of above
type GetTypeMapped<T extends RecordOfTypes, P> = { [K in keyof T]: GetType<T[K], P> };

// Creates an object type
declare function MakeObject<T extends RecordOfTypes>(object: T): TObject<T>;
interface TObject<T extends RecordOfTypes> {
    objectProps: unknown;
    myType: GetTypeMapped<T, this['objectProps']>;
}

// Any dictionary from string -> statically-typed type
type RecordOfTypes = Record<string, HasType>;
```

This example defeats the existing "are we making progress" heuristics because of
```
    myType: GetTypeMapped<T, this['objectProps']>;
```
When computing `Output.level1.level2`, this is computing `this['objectProps']['objectProps']['objectProps']`, which looks a heck of a lot like the infinite series of `next`s in `RabbitHole`.

**Comment by sinclairzx81** [comment]
@RyanCavanaugh 

> Reduced and annotated example (apologies for renaming but these make more sense to me and avoid conflicting with certain terms we use internally):

Hey no problem. Also, thank you for taking the time to investigate this issue (and for the test PR), that was very much appreciated. If there is anything that can be done library side to help work within TS depth limits here, please let me know.

**Comment by fatcerberus** [comment]
> The PR tanked xstate performance; we'll need something smarter.

Didn't @ahejlsberg recently author a PR with smarter logic for this?

**Comment by sinclairzx81** [comment]
Related https://github.com/microsoft/TypeScript/pull/55638

**Comment by fatcerberus** [comment]
@sinclairzx81 That was the PR I was thinking of, thanks!

**Comment by webstrand** [comment]
Could we perhaps explicitly mark these cases instead, if detecting them automatically proves infeasible?

**Comment by webstrand** [comment]
I mean an intrinsic `Converges<T>` (or a keyword) which _may_ loop infinitely if poorly written. I recognize that detecting these cases automatically may be impossible, but rather than an esoteric "you must write your recursive types in a particular pattern that typescript recognizes, but is only known to the few who frequent issues" mark it explicitly.

**Comment by ahejlsberg** [comment]
With #55638 we got pretty close to fixing the issue. However, use of the non-homomorphic `Pick` and the `Evaluate` conditional causes the compiler to loose track of the symbol of the original literal type. Once that happens, all of the transformed types have the same symbol and we stop relating them after three levels. The code works if I remove the use of `Evaluate` and change to a homomorphic definition of `Pick`:

```ts
type Pick<T, K extends keyof T> = { [P in keyof T as P & K]: T[P] };
```

I'll keep thinking about this, but it's a really tricky issue to solve.

**Comment by sinclairzx81** [comment]
Possible library level workaround (but not sure how well this would scale)

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/PTAEFpK6dv4YpyUIFAlAdQPYCcBrAQzxwFcA7AEwC5QALAF0YAcBnGkAcwEtH6yAIwB0AYxwBbYBJ6jSbHADNGwACoBPFgFMAynJ4sVPNmzJa2wAKwA2AIwBmABwYwqN+4+ev4NGi0APFnxGUEZNLVAAJS1RMjw2HgA3CIBeUB1GIkZZAB4w7SUomLiE5IA+P0Dg0HEKNhDo2Pik1NANbWFGkpaAClV6Y1AUsrbw4QB5QQArGMYegG9QIjp2rQnp2YXQQRWBtlAAXwBKQ6Oj3wCgvBD8iIBJChYyELSMrNzbwoenxgrL6tq9VA32eQ1GHUmM1Ec3maFAoAANlpkgjbCsxpDNrD4fCkSiAEzoiEbaELOE43HIrQI+xEtaY0nYik4vHUgAsdPWUJh5OZOMUOBwnIyeB4FC4PSOABpeczjrL4fKKUrTmh5ZUrjdwqBxs8fmC3tlRHlwoVdaxnn8qtcajg6iFzfq0qsuVjyazUZyGTyKR7CeD6SSfcyPbSA67GQqWVSERzw96yXyKQKheGRWKJdKoxTBCRhYxReLJTKk6q+SrFedK2rzmgqDEESQIoCQnhis1knQuh2tABuUBtpqlNZEYQiUfjsfCCfTqcz+dz0CYHAEXyKSjQnh20AsUiCJESd6iABiG+ydtsPTFPzoIMYAG0ALpHOiO55P0BMtuMOIUUDX55+2XVcDjQdcKE3bddxwfctEPI1Twg88KHxHJzVAAJGC0ah9jfB9HzKK9HmeW9iPwl8dVBL8tB-PA-wAxggLAFc1V8TBvA4ziuNQFx0kyI1QF47jhJE4SNWqW5QAABVIbRrh4cwAAkiDYfpjFWHJVAw-wsJwtoZJwOTsnMEY0kWcl7wAaX-P8CC0dRClUR86GU1S9g01QrII7TdKofYCzMUAAH5QFsUA6AABkEw57zshzFDaR8fOwvybMULQ8CiaL4RCsLMJS-ZImynEQoCiIFToRQiARNhypxOgKCpPALmtLVtFAVy1NU8JNOSvTVD0eg4KIUzyS0-L+q64LQjwQK6DGvrUtUb0cjFdLMqkkYQoMoyFLYTr3J6zbwuyqqaq0FrNVCbVDVycadIKtpBuGqVpMW-ZKAICgcAAdwoD80ifV7XLBA71J61QylMjqVPembAvJEK+lAAAyT8dxIIgJA4N7jnvAByeoj3xpKlzAQdumSDH+HJFYCaJo0SdLZmcUwRQxT4CIWCyeg2NcUSBcFtxeIG0QhsPIShal6XYHEm0xSwvAqtECJRfFohP3JBnZDoT7vr+8lubwLGcb137-sfVj2Jlm3pZF71Jdtp3RLltrVZ2jLjP2NJGnwKgcnqQsuFetXhqtK6FYy5XVZW+7fP2VQPfkky4dDw9NfhW46HxmDuXxrX+J19GrJs0A4sc5y+KPTSvNe-hjAJo2TZJkZQNA63nc7riRfTcVHa7gfPFdmzFejp6CwzVPnvTpks9AQmJ-FfP4W10Q6EDjMrf5wed+FzBVC6-vd+P9B-nlihR6IFW2imialunjXZ-CbP67YZfQFXuhX8bzHsZJ+8IqW3btvE+oDZb727MOI+YCYHD0kqoSBLRohUDICrXqd8E4P2hrdY0qhXr3gQe2YcyDUFaE0mUAiw9I5KyvqrRByR0EPX6lgqeYthoZ2utobOFMezv0-m0ehWgSFoMhlvCAMCJGQBFuEaBkjj7D3rKIRsbZQAUCxuYbm19VgcLPiEcCkE-yCMYfHJ6bDDyEVENVBEuZRAEDoD0V+qwVhdROMMNoFFCFDhaOQ8kujQD6OQjqIMxjHqJ1kp7PahFoK7XMCsDxscKjwj8QErcf5e6ZhWOk1iaAgA)

```typescript
export type PropertiesHasThisType<T extends TProperties> = { 
  [K in keyof T]: HasThisType<T[K]> extends true ? 1 : 0 
 }[keyof T] extends infer R 
   ? 1 extends R 
     ? true 
     : false 
   : never

export type HasThisType<T extends TSchema> =
  T extends TThis ? true : 
  T extends TObject<infer P> ? PropertiesHasThisType<P> : 
  false

export type Static<T extends TSchema, P extends unknown[] = [], Has = HasThisType<T>> = Has extends true 
  ? (T & { params: P })['static']  // recursive path
  : T['static']                    // finite path
  ```

**Comment by ahejlsberg** [comment]
Actually, it appears possible to fix this by improving our logic for tracking deeply nested types in relationships. I will put up a PR.

**Comment by sinclairzx81** [comment]
@ahejlsberg This PR is amazing. Thank you so much!
